c++/src/ColumnWriter.cc

```diff
@@ -468,25 +472,1099 @@ namespace orc {
+    encodings.push_back(encoding);
+  }
+
+  void BooleanColumnWriter::recordPosition() const {
+    ColumnWriter::recordPosition();
+    rleEncoder->recordPosition(rowIndexPosition.get());
+  }
+
+  class DoubleColumnWriter : public ColumnWriter {
+  public:
+    DoubleColumnWriter(const Type& type,
+                       const StreamsFactory& factory,
+                       const WriterOptions& options,
+                       bool isFloat);
+
+    virtual void add(ColumnVectorBatch& rowBatch,
+                     uint64_t offset,
+                     uint64_t numValues) override;
+
+    virtual void flush(std::vector<proto::Stream>& streams) override;
+
+    virtual uint64_t getEstimatedSize() const override;
+
+    virtual void getColumnEncoding(
+        std::vector<proto::ColumnEncoding>& encodings) const override;
+
+    virtual void recordPosition() const override;
+
+  private:
+    bool isFloat;
+    std::unique_ptr<AppendOnlyBufferedStream> dataStream;
+    DataBuffer<char> buffer;
+  };
+
+  DoubleColumnWriter::DoubleColumnWriter(
+                          const Type& type,
+                          const StreamsFactory& factory,
+                          const WriterOptions& options,
+                          bool isFloatType) :
+                              ColumnWriter(type, factory, options),
+                              isFloat(isFloatType),
+                              buffer(*options.getMemoryPool()) {
+    dataStream.reset(new AppendOnlyBufferedStream(
+                             factory.createStream(proto::Stream_Kind_DATA)));
+    buffer.resize(isFloat ? 4 : 8);
+
+    if (enableIndex) {
+      recordPosition();
+    }
+  }
+
+  void DoubleColumnWriter::add(ColumnVectorBatch& rowBatch,
+                               uint64_t offset,
+                               uint64_t numValues) {
+    ColumnWriter::add(rowBatch, offset, numValues);
+
+    const DoubleVectorBatch& dblBatch =
+                             dynamic_cast<const DoubleVectorBatch&>(rowBatch);
+
+    const double* doubleData = dblBatch.data.data() + offset;
+    const char* notNull = dblBatch.hasNulls ?
+                          dblBatch.notNull.data() + offset : nullptr;
+
+    size_t bytes = isFloat ? 4 : 8;
+    char* data = buffer.data();
+
+    for (uint64_t i = 0; i < numValues; ++i) {
+      if (!notNull || notNull[i]) {
+        if (isFloat) {
+          // to avoid float-double cast
+          const int32_t* intBits =
+            reinterpret_cast<const int32_t*>(&static_cast<const float&>(
+              doubleData[i]));
+          for (size_t j = 0; j < bytes; ++j) {
+            data[j] = static_cast<char>(((*intBits) >> (8 * j)) & 0xff);
+          }
+        } else {
+          const int64_t* intBits =
+            reinterpret_cast<const int64_t*>(&(doubleData[i]));
+          for (size_t j = 0; j < bytes; ++j) {
+            data[j] = static_cast<char>(((*intBits) >> (8 * j)) & 0xff);
+          }
```

Comment:
I think the intent of this and the block on line 731 would be a lot more clear if you made a helper function that took input and output pointers and a type width to handle an endian agnostic copy.

Reply:
Done